sql/plan/alter_table.go

```diff
@@ -1121,6 +1134,153 @@ func (i *modifyColumnIter) Close(context *sql.Context) error {
+	for i, col := range newSch {
+		if col.AutoIncrement {
+			projections[i], err = expression.NewAutoIncrementForColumn(ctx, rwt, col, projections[i])
+			if err != nil {
+				return false, err
+			}
+		}
+	}
+
+	oldPkSchema, newPkSchema := sql.SchemaToPrimaryKeySchema(rwt, rwt.Schema()), sql.SchemaToPrimaryKeySchema(rwt, newSch)
+
+	// TODO: codify rewrite requirements
+	rewriteRequired := true
+
+	rewriteRequested := rwt.ShouldRewriteTable(ctx, oldPkSchema, newPkSchema, i.m.column)
+	if !rewriteRequired && !rewriteRequested {
+		return false, nil
+	}
+
+	inserter, err := rwt.RewriteInserter(ctx, newPkSchema)
+	if err != nil {
+		return false, err
+	}
+
+	partitions, err := rwt.Partitions(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	rowIter := sql.NewTableRowIter(ctx, rwt, partitions)
+
+	for {
+		r, err := rowIter.Next(ctx)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return false, err
+		}
+
+		newRow, err := projectRowWithTypes(ctx, newSch, projections, r)
+		if err != nil {
+			return false, err
+		}
+
+		err = inserter.Insert(ctx, newRow)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// TODO: move this into iter.close, probably
+	err = inserter.Close(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// projectRowWithTypes projects the row given with the projections given and additionally converts them to the
+// corresponding types found in the schema given, using the standard type conversion logic.
+func projectRowWithTypes(ctx *sql.Context, sch sql.Schema, projections []sql.Expression, r sql.Row) (sql.Row, error) {
+	newRow, err := ProjectRow(ctx, projections, r)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range newRow {
+		newRow[i], err = sch[i].Type.Convert(newRow[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return newRow, nil
+}
+
+// modifyColumnInSchema modifies the given column in given schema and returns the new schema, along with a set of
+// projections to adapt the old schema to the new one.
+func modifyColumnInSchema(schema sql.Schema, name string, column *sql.Column, order *sql.ColumnOrder) (sql.Schema, []sql.Expression, error) {
+	currIdx := schema.IndexOf(name, column.Source)
+	if currIdx < 0 {
+		// Should be checked in the analyzer already
+		return nil, nil, sql.ErrTableColumnNotFound.New(column.Source, name)
+	}
+
+	// Primary key-ness isn't included in the column description as part of the ALTER statement, preserve it
+	if schema[currIdx].PrimaryKey {
+		column.PrimaryKey = true
+	}
+
+	newIdx := currIdx
+	if order != nil && len(order.AfterColumn) > 0 {
+		newIdx = schema.IndexOf(order.AfterColumn, column.Source)
+		if newIdx == -1 {
+			// Should be checked in the analyzer already
+			return nil, nil, sql.ErrTableColumnNotFound.New(column.Source, order.AfterColumn)
+		}
+	} else if order != nil && order.First {
+		newIdx = 0
+	}
+
+	// establish a map from old column index to new column index
+	oldToNewIdxMapping := make(map[int]int)
```

Comment:
this is what i did for ordinals, you said it was overly clever and I should just create a name->pos map on the old schema and use it to create the oldPos->newPos map

Reply:
I might have had a point